diff -up ./nss/cmd/Makefile.skipthem ./nss/cmd/Makefile
--- ./nss/cmd/Makefile.skipem	2016-06-24 10:10:38.143165159 -0700
+++ ./nss/cmd/Makefile	2016-06-24 10:13:08.566457400 -0700
@@ -17,7 +17,11 @@ endif
 ifeq ($(NSS_BUILD_WITHOUT_SOFTOKEN),1)
 BLTEST_SRCDIR =
 FIPSTEST_SRCDIR =
+ifeq ($(NSS_BLTEST_NOT_AVAILABLE),1)
+SHLIBSIGN_SRCDIR = shlibsign
+else
 SHLIBSIGN_SRCDIR =
+endif
 else
 BLTEST_SRCDIR = bltest
 FIPSTEST_SRCDIR = fipstest
