diff -up ./nss/external_tests/manifest.mn.skip_util_gtest ./nss/external_tests/manifest.mn
--- ./nss/external_tests/manifest.mn.skip_util_gtest	2016-05-21 21:34:56.156346633 -0700
+++ ./nss/external_tests/manifest.mn	2016-05-21 21:35:23.408854282 -0700
@@ -8,7 +8,6 @@ DEPTH      = ..
 DIRS = \
 	google_test \
         der_gtest \
-	util_gtest \
         pk11_gtest \
         ssl_gtest \
 	$(NULL)
