diff -up ./mozilla/security/nss/lib/ckfw/manifest.mn.libpem ./mozilla/security/nss/lib/ckfw/manifest.mn
--- ./mozilla/security/nss/lib/ckfw/manifest.mn.libpem	2007-08-15 17:29:52.000000000 +0200
+++ ./mozilla/security/nss/lib/ckfw/manifest.mn	2009-01-23 05:49:59.000000000 +0100
@@ -38,7 +38,7 @@ MANIFEST_CVS_ID = "@(#) $RCSfile: manife
 
 CORE_DEPTH = ../../..
 
-DIRS = builtins 
+DIRS = builtins pem
 
 PRIVATE_EXPORTS = \
 	ck.h		  \
