--- ./lib/nss/config.mk.ignore_system_policy	2016-06-08 20:25:58.841775044 -0700
+++ ./lib/nss/config.mk	2016-06-08 20:27:06.507526784 -0700
@@ -101,10 +101,12 @@
 # Export 'mktemp' to be backward compatible with NSS 3.2.x and 3.3.x
 # but do not put it in the import library.  See bug 142575.
 DEFINES += -DWIN32_NSS3_DLL_COMPAT
 DLLFLAGS += -EXPORT:mktemp=nss_mktemp,PRIVATE
 endif
 endif
 
 ifdef POLICY_FILE
+ifndef NSS_IGNORE_SYSTEM_POLICY
 DEFINES += -DPOLICY_FILE=\"$(POLICY_FILE)\" -DPOLICY_PATH=\"$(POLICY_PATH)\"
 endif
+endif
