diff -up nss/tests/chains/scenarios/scenarios.nocsptest nss/tests/chains/scenarios/scenarios
--- nss/tests/chains/scenarios/scenarios.nocsptest	2013-07-02 13:16:31.833623231 -0700
+++ nss/tests/chains/scenarios/scenarios	2013-07-02 13:17:02.465910005 -0700
@@ -48,7 +48,6 @@ bridgewithaia.cfg
 bridgewithhalfaia.cfg
 bridgewithpolicyextensionandmapping.cfg
 realcerts.cfg
-dsa.cfg
 revoc.cfg
 ocsp.cfg
 crldp.cfg
