--- ./lib/nss/config.mk.ignore_system_policy	2016-06-08 09:50:41.156814910 -0700
+++ ./lib/nss/config.mk	2016-06-08 09:53:09.874120131 -0700
@@ -100,11 +100,13 @@
 ifndef NS_USE_GCC
 # Export 'mktemp' to be backward compatible with NSS 3.2.x and 3.3.x
 # but do not put it in the import library.  See bug 142575.
 DEFINES += -DWIN32_NSS3_DLL_COMPAT
 DLLFLAGS += -EXPORT:mktemp=nss_mktemp,PRIVATE
 endif
 endif
 
+ifndef NSS_IGNORE_SYSTEM_POLICY
 ifdef POLICY_FILE
 DEFINES += -DPOLICY_FILE=\"$(POLICY_FILE)\" -DPOLICY_PATH=\"$(POLICY_PATH)\"
 endif
+endif
