Index: mozilla/security/nss/lib/ckfw/manifest.mn
===================================================================
RCS file: /cvsroot/mozilla/security/nss/lib/ckfw/manifest.mn,v
retrieving revision 1.9
diff -u -r1.9 manifest.mn
--- mozilla/security/nss/lib/ckfw/manifest.mn   16 Dec 2005 00:48:01 -0000      1.9
+++ mozilla/security/nss/lib/ckfw/manifest.mn   31 Aug 2007 17:58:23 -0000
@@ -38,7 +38,7 @@
 
 CORE_DEPTH = ../../..
 
-DIRS = builtins 
+DIRS = builtins pem
 
 PRIVATE_EXPORTS = \
        ck.h              \
