diff -up ./lib/ssl/ssl3exthandle.c.rebuild_errors ./lib/ssl/ssl3exthandle.c
--- ./lib/ssl/ssl3exthandle.c.rebuild_errors	2023-01-24 09:26:36.520183263 -0800
+++ ./lib/ssl/ssl3exthandle.c	2023-01-24 09:27:07.715379228 -0800
@@ -201,7 +201,7 @@ ssl3_FreeSniNameArray(TLSExtensionData *
  * Clients sends a filled in session ticket if one is available, and otherwise
  * sends an empty ticket.  Servers always send empty tickets.
  */
-PRInt32
+SECStatus
 ssl3_ClientSendSessionTicketXtn(const sslSocket *ss, TLSExtensionData *xtnData,
                                 sslBuffer *buf, PRBool *added)
 {
diff -up ./lib/ssl/sslsnce.c.rebuild_errors ./lib/ssl/sslsnce.c
--- ./lib/ssl/sslsnce.c.rebuild_errors	2023-01-24 09:44:52.714977837 -0800
+++ ./lib/ssl/sslsnce.c	2023-01-24 09:46:20.993510435 -0800
@@ -1820,7 +1820,7 @@ ssl_GetSelfEncryptKeyPair(SECKEYPublicKe
     return SECSuccess;
 }
 
-static PRBool
+static SECStatus
 ssl_GenerateSelfEncryptKeys(void *pwArg, PRUint8 *keyName,
                             PK11SymKey **aesKey, PK11SymKey **macKey);
 
