diff -ruNp mozilla.orig/security/nss/lib/ckfw/pem/pobject.c mozilla/security/nss/lib/ckfw/pem/pobject.c
--- mozilla.orig/security/nss/lib/ckfw/pem/pobject.c	2009-02-28 15:03:36.719964516 +0100
+++ mozilla/security/nss/lib/ckfw/pem/pobject.c	2009-02-28 15:03:55.964074075 +0100
@@ -987,10 +987,7 @@ pem_CreateObject
     char *ivstring = NULL;
 
     count = pem_nobjs;
-    if (count > 0)
-        size = ((count / PEM_ITEM_CHUNK) + 1) * PEM_ITEM_CHUNK;
-    else
-        size = 0;
+    size = (count / PEM_ITEM_CHUNK) * PEM_ITEM_CHUNK;
 
     /*
      * only create token objects
