diff -up ./lib/freebl/blinit.c.disable_hw_ppc ./lib/freebl/blinit.c
--- ./lib/freebl/blinit.c.disable_hw_ppc	2021-05-27 18:04:59.754657701 -0700
+++ ./lib/freebl/blinit.c	2021-05-27 18:07:02.756397733 -0700
@@ -502,6 +502,7 @@ CheckPPCSupport()
     char *disable_hw_crypto = PR_GetEnvSecure("NSS_DISABLE_PPC_GHASH");
 
     unsigned long hwcaps = 0;
+#ifdef notdef
 #if defined(__linux__)
 #if __has_include(<sys/auxv.h>)
     hwcaps = getauxval(AT_HWCAP2);
@@ -516,6 +517,7 @@ CheckPPCSupport()
     sysctlbyname("hw.cpu_features2", &hwcaps, &len, NULL, 0);
 #endif
 #endif
+#endif
 
     ppc_crypto_support_ = hwcaps & PPC_FEATURE2_VEC_CRYPTO && disable_hw_crypto == NULL;
 }
