--- ./lib/ssl/Makefile.allow_keylogfile	2016-06-08 21:54:29.504328764 -0700
+++ ./lib/ssl/Makefile	2016-06-08 22:03:57.061313047 -0700
@@ -34,19 +34,20 @@
 else
 ifeq ($(OS_TARGET),OS2)
 CSRCS	+= os2_err.c
 else
 CSRCS	+= unix_err.c
 endif
 endif
 
-# Enable key logging by default in debug builds, but not opt builds.
-# Logging still needs to be enabled at runtime through env vars.
-NSS_ALLOW_SSLKEYLOGFILE ?= $(if $(BUILD_OPT),0,1)
+# For Fedora stable branch compatibility, i.e f23 and f22
+# Enable key logging by default in debug builds and opt builds.
+# Logging doesn't need to be enabled at runtime through env vars.
+NSS_ALLOW_SSLKEYLOGFILE = 1
 ifeq (1,$(NSS_ALLOW_SSLKEYLOGFILE))
 DEFINES += -DNSS_ALLOW_SSLKEYLOGFILE=1
 endif
 
 #######################################################################
 # (5) Execute "global" rules. (OPTIONAL)                              #
 #######################################################################
 
