diff -up ./nss/lib/ckfw/pem/pobject.c.pemAll ./nss/lib/ckfw/pem/pobject.c
--- ./nss/lib/ckfw/pem/pobject.c.pemAll	2014-01-23 06:28:18.000000000 -0800
+++ ./nss/lib/ckfw/pem/pobject.c	2015-11-11 15:15:16.581569576 -0800
@@ -630,6 +630,8 @@ pem_DestroyInternalObject
         if (io->u.key.ivstring)
             free(io->u.key.ivstring);
         break;
+    case pemAll:
+        return;
     }
 
     if (NULL != gobj)
@@ -1044,7 +1046,7 @@ pem_CreateObject
     int nobjs = 0;
     int i;
     int objid;
-    pemToken *token;
+    /*pemToken *token = NULL;*/
     int cipher;
     char *ivstring = NULL;
     pemInternalObject *listObj = NULL;
@@ -1073,7 +1075,7 @@ pem_CreateObject
     }
     slotID = nssCKFWSlot_GetSlotID(fwSlot);
 
-    token = (pemToken *) mdToken->etc;
+    /*token = (pemToken *) mdToken->etc;*/
 
     /*
      * only create keys and certs.
