diff -up ./nss/lib/ssl/sslsock.c.transitional ./nss/lib/ssl/sslsock.c
--- ./nss/lib/ssl/sslsock.c.transitional	2016-03-05 08:54:13.871412639 -0800
+++ ./nss/lib/ssl/sslsock.c	2016-03-05 09:00:27.721889811 -0800
@@ -77,7 +77,7 @@ static sslOptions ssl_defaults = {
     PR_FALSE,                /* noLocks            */
     PR_FALSE,                /* enableSessionTickets */
     PR_FALSE,                /* enableDeflate      */
-    2,                       /* enableRenegotiation (default: requires extension) */
+    3,                       /* enableRenegotiation (default: transitional) */
     PR_FALSE,                /* requireSafeNegotiation */
     PR_FALSE,                /* enableFalseStart   */
     PR_TRUE,                 /* cbcRandomIV        */
