diff -up ./mozilla/security/nss/cmd/addbuiltin/addbuiltin.c.terminalrecord ./mozilla/security/nss/cmd/addbuiltin/addbuiltin.c
--- ./mozilla/security/nss/cmd/addbuiltin/addbuiltin.c.terminalrecord	2011-12-04 21:15:47.660672521 -0800
+++ ./mozilla/security/nss/cmd/addbuiltin/addbuiltin.c	2011-12-04 21:16:39.491674974 -0800
@@ -52,6 +52,14 @@
 #include <io.h>
 #endif
 
+/* Defined in ./mozilla/dist/public/nss/certdb.h which was included
+ * and also in ./mozilla/security/nss/lib/softoken/legacydb/pcertt.h
+ * but invisible here for some reason
+ */
+#ifndef CERTDB_TERMINAL_RECORD
+#define CERTDB_TERMINAL_RECORD	(1<<0)
+#endif
+
 void dumpbytes(unsigned char *buf, int len)
 {
     int i;
diff -up ./mozilla/security/nss/lib/certhigh/certvfy.c.terminalrecord ./mozilla/security/nss/lib/certhigh/certvfy.c
--- ./mozilla/security/nss/lib/certhigh/certvfy.c.terminalrecord	2011-09-13 17:28:47.000000000 -0700
+++ ./mozilla/security/nss/lib/certhigh/certvfy.c	2011-12-04 21:14:01.683678605 -0800
@@ -56,6 +56,14 @@
 #include "pki3hack.h"
 #include "base.h"
 
+/* Defined in ./mozilla/dist/public/nss/certdb.h which was included
+ * and also in ./mozilla/security/nss/lib/softoken/legacydb/pcertt.h
+ * but invisible here for some reason
+ */
+#ifndef CERTDB_TERMINAL_RECORD
+#define CERTDB_TERMINAL_RECORD	(1<<0)
+#endif
+
 /*
  * Check the validity times of a certificate
  */
diff -up ./mozilla/security/nss/lib/libpkix/pkix_pl_nss/pki/pkix_pl_cert.c.terminalrecord ./mozilla/security/nss/lib/libpkix/pkix_pl_nss/pki/pkix_pl_cert.c
--- ./mozilla/security/nss/lib/libpkix/pkix_pl_nss/pki/pkix_pl_cert.c.terminalrecord	2011-09-14 16:16:16.000000000 -0700
+++ ./mozilla/security/nss/lib/libpkix/pkix_pl_nss/pki/pkix_pl_cert.c	2011-12-04 21:14:01.684678605 -0800
@@ -43,6 +43,14 @@
 
 #include "pkix_pl_cert.h"
 
+/* Defined in ./mozilla/dist/public/nss/certdb.h which was included
+ * and also in ./mozilla/security/nss/lib/softoken/legacydb/pcertt.h
+ * but invisible here for some reason
+ */
+#ifndef CERTDB_TERMINAL_RECORD
+#define CERTDB_TERMINAL_RECORD	(1<<0)
+#endif
+
 extern PKIX_PL_HashTable *cachedCertSigTable;
 
 /* --Private-Cert-Functions------------------------------------- */
diff -up ./mozilla/security/nss/lib/pk11wrap/pk11nobj.c.terminalrecord ./mozilla/security/nss/lib/pk11wrap/pk11nobj.c
--- ./mozilla/security/nss/lib/pk11wrap/pk11nobj.c.terminalrecord	2011-04-12 17:10:26.000000000 -0700
+++ ./mozilla/security/nss/lib/pk11wrap/pk11nobj.c	2011-12-04 21:14:01.685678605 -0800
@@ -62,6 +62,14 @@
 #include "pki.h"
 #include "pkim.h" 
 
+/* Defined in ./mozilla/dist/public/nss/certdb.h which was included
+ * and also in ./mozilla/security/nss/lib/softoken/legacydb/pcertt.h
+ * but invisible here for some reason
+ */
+#ifndef CERTDB_TERMINAL_RECORD
+#define CERTDB_TERMINAL_RECORD	(1<<0)
+#endif
+
 extern const NSSError NSS_ERROR_NOT_FOUND;
 
 CK_TRUST
diff -up ./mozilla/security/nss/lib/pki/pki3hack.c.terminalrecord ./mozilla/security/nss/lib/pki/pki3hack.c
--- ./mozilla/security/nss/lib/pki/pki3hack.c.terminalrecord	2011-04-12 17:10:26.000000000 -0700
+++ ./mozilla/security/nss/lib/pki/pki3hack.c	2011-12-04 21:14:01.686678605 -0800
@@ -76,6 +76,14 @@ static const char CVS_ID[] = "@(#) $RCSf
 #include "secmod.h"
 #include "nssrwlk.h"
 
+/* Defined in ./mozilla/dist/public/nss/certdb.h which was included
+ * and also in ./mozilla/security/nss/lib/softoken/legacydb/pcertt.h
+ * but invisible here for some reason
+ */
+#ifndef CERTDB_TERMINAL_RECORD
+#define CERTDB_TERMINAL_RECORD	(1<<0)
+#endif
+
 NSSTrustDomain *g_default_trust_domain = NULL;
 
 NSSCryptoContext *g_default_crypto_context = NULL;
